### DISCUSSION TITLE
[PROPOSAL] Fix HDF4 library naming / organizational scheme

### DISCUSSION THREAD

**Discussion by derobins** [HDFGroup/hdf4#409] [Proposed Change]
### **PROBLEM**

The Autotools and CMake have very different ideas about how to name and organize their output.

From the Autotools (shared/Java/XDR - Fortran built separately as static)
bin
```
gif2hdf  h4redeploy  hdf2gif   hdf8to24  hdfed      hdfls    hdftopal  hdfunpac  hdp      jpeg2hdf  ncgen     r8tohdf   vmake
h4cc     hdf24to8    hdf2jpeg  hdfcomp   hdfimport  hdfpack  hdftor8   hdiff     hrepack  ncdump    paltohdf  ristosds  vshow
```
lib
```
jarhdf-4.2.17-1.jar  libdf.so        libhdf4.settings  libhdf_java.so        libmfhdf.a   libmfhdf.so.0      libxdr.la    libxdr.so.0.0.0
libdf.a              libdf.so.0      libhdf_java.a     libhdf_java.so.0      libmfhdf.la  libmfhdf.so.0.0.0  libxdr.so
libdf.la             libdf.so.0.0.0  libhdf_java.la    libhdf_java.so.0.0.0  libmfhdf.so  libxdr.a           libxdr.so.0
```

From CMake (shared/Fortran/Java/XDR)
bin
```
h4cc   hdfed-shared      hdfls         hdiff-shared  hrepack         ncdump-shared
h4fc   hdfimport         hdfls-shared  hdp           hrepack-shared  ncgen
hdfed  hdfimport-shared  hdiff         hdp-shared    ncdump          ncgen-shared
```
lib
```
jarhdf-4.2.17.jar        libhdf_fortran.a          libhdf.so             libmfhdf_fcstub.so.4.17.2   libmfhdf.so.4.17.2    slf4j-nop-1.7.33.jar
libhdf4.settings         libhdf_fortran.so         libhdf.so.4           libmfhdf_fortran.a          libxdr.a              slf4j-simple-1.7.33.jar
libhdf.a                 libhdf_fortran.so.4       libhdf.so.4.17.2      libmfhdf_fortran.so         libxdr.so
libhdf_fcstub.a          libhdf_fortran.so.4.17.2  libmfhdf.a            libmfhdf_fortran.so.4       libxdr.so.4
libhdf_fcstub.so         libhdf_java.so            libmfhdf_fcstub.a     libmfhdf_fortran.so.4.17.2  libxdr.so.4.17.2
libhdf_fcstub.so.4       libhdf_java.so.4          libmfhdf_fcstub.so    libmfhdf.so                 pkgconfig
libhdf_fcstub.so.4.17.2  libhdf_java.so.4.17.2     libmfhdf_fcstub.so.4  libmfhdf.so.4               slf4j-api-1.7.33.jar
```
Note the differences:
* CMake adds -shared to tools linked against the shared library
* CMake creates `libhdf` vs Autotools `libdf`
* Autotools sonames differ
* CMake puts the Fortran wrappers in separate libraries

### **PROPOSED FIX**

* Drop "-shared" from shared tools
* Uniform names (how do we handle libhdf vs libdf? - maybe use libhdf and symlink libdf)
* The Autotools should create separate Fortran libraries like CMake and support shared libraries

### **HOW WOULD USERS BE AFFECTED?**

**PRO**
➕ CMake and Autotools products would have identical names
➕ Autotools Fortran output would be more modern, more like HDF5 (support shared, easier to bring improvements over)

**CON**
➖ Some users may be unhappy with the new names

**Comment by stevengj** [comment]
Changing the name of the library (from libdf to libhdf) will break every single HDF4-based build script that has ever existed… this isn't something to do lightly, especially since HDF4-based projects are largely legacy code, which may not be actively maintained.
